dashboard/app/models/level_group.rb

```diff
@@ -87,4 +88,112 @@ def pages
+      case sublevel
+      when TextMatch, FreeResponse
+        student_result = sublevel_response["result"]
+        sublevel_result[:result] = student_result
+        sublevel_result[:type] = "free_response"
+      when Multi
+        student_result = sublevel_response["result"].split(",").sort.join(",")
+        # unless unsubmitted
+        unless student_result == "-1"
+          answer_text = sublevel.properties.try(:[], "answers").try(:[], student_result.to_i).try(:[], "text")
+          sublevel_result[:result_text] = answer_text
+          # Convert "0,1,3" to "A, B, D" for teacher-friendly viewing
+          sublevel_result[:result] = student_result.split(',').map{ |k| Multi.value_to_letter(k.to_i) }.join(', ')
+          sublevel_result[:type] = "multi"
+        end
+      end
+    end
+    sublevel_result
+  end
+
+  # Surveys: Returns all survey results given a student and a script level of type LevelGroup.
+  def self.get_survey_results_for_student(student, script_level)
+    student_results = {}
+
+    # Skip student if they haven't submitted for this LevelGroup.
+    user_level = student.user_level_for(script_level, script_level.level)
+    return unless user_level.try(:submitted)
+
+    last_attempt = student.last_attempt(script_level.level)
+    response = last_attempt.try(:level_source).try(:data)
+    return unless response
+
+    response_parsed = JSON.parse(response)
+
+    # Go through each sublevel
+    script_level.level.levels.each_with_index do |sublevel, sublevel_index|
+      # If it's the first time we're saving a result for this sublevel, then
+      # create the full entry.
+      unless student_results[sublevel_index]
+        question_text = sublevel.properties.try(:[], "questions").try(:[], 0).try(:[], "text") ||
+                        sublevel.properties.try(:[], "markdown_instructions")
+        student_results[sublevel_index] = {question: question_text, results: []}
+      end
+
+      sublevel_response = response_parsed[sublevel.id.to_s]
+
+      student_results[sublevel_index][:results] << get_sublevel_result(sublevel, sublevel_response)
+    end
+
+    student_results
+  end
+
+  # Surveys: Returns all anonymized survey results, given a script and a section.
+  def self.get_survey_results(script, section)
+    level_group_script_levels = script.script_levels.includes(:levels).where('levels.type' => LevelGroup)
+
+    # Go through each anonymous long-assessment LevelGroup.
+    level_group_script_levels.map do |script_level|
+      next unless script_level.long_assessment?
+      next unless script_level.anonymous?
+
+      levelgroup_results = {}
+
+      # Go through each student who has submitted a response to it.
+      section.students.each do |student|
+        student_result = get_survey_results_for_student(student, script_level)
+        if student_result
+          student_result.each do |key, value|
+            levelgroup_results[key] = {results: []} unless levelgroup_results[key]
```

Comment:
Not sure that works because we are adding an entry once we know the key (aka sublevel_index), and we don't know that at the time of initializing levelgroup_results.